modules/scrollback/index.js

```diff
@@ -110,108 +111,393 @@ module.exports.formatLine = function(line) {
   }
 
-  var i;
+  // Parse input into tokens
+  while(i < input.length) {
+    if(state === 'nick') {
+      result = parseNick(input, i);
+
+      if(i != result.index) { // Got a nick
+        i = result.index;
+        spec.nicks.push(result.result);
+      }
+
+    }
+    else if(state == 'regex') {
+      result = parseRegex(input, i);
+
+      if(result.error !== undefined) {
+        error = result.error;
+        state = 'error';
+      }
+      else {
+        if(i != result.index) { // Got a regex
+          i = result.index;
+          spec.regexes.push(result.result);
+        }
+      }
+    }
+    else if(state == 'lines') {
+      result = parseLines(input, i);
 
-  for(i=0;i<parts.length;i++) {
-    // TODO, there's some serious duplication going on here
-    if(/^(-|\d)/.test(parts[i])) {
-      // Starts with a - or number, can't be a nick
-      linesToGet.push(rangeParser.parse(parts[i]));
-    } else if(/^\/.*\/$/.test(parts[i])) {
-      // sandwitched between '/'s, it's a regex
-      try {
-        var regex = new RegExp(parts[i].substring(1, parts[i].length - 1));
-        linesToGet.push(regex);
-      } catch(ex) {
-        return cb("Could not parse regex: " + parts[i] + ", " + ex.toString());
+      if(result.error !== undefined) {
+        error = result.error;
+        state = 'error';
       }
-    } else {
-      // This is a nick. If the next one is a number then it's for this nick
-      var lineObj = {
-        from: parts[i]
-      };
-      i++;
-      if(/^(-|\d)/.test(parts[i])) {
-        lineObj.lines = rangeParser.parse(parts[i]);
-      } else if(/^\/.*\/$/.test(parts[i])) {
-        // sandwitched between '/'s, it's a regex
-        try {
-          var regex = new RegExp(parts[i].substring(1, parts[i].length - 1));
-          lineObj.regex = regex;
-        } catch(ex) {
-          return cb("Could not parse regex: " + parts[i] + ", " + ex.toString());
+      else {
+        if(i != result.index) { // Got some lines
+          i = result.index;
+          spec.lines = spec.lines.concat(result.result);
         }
-      } else {
-        lineObj.lines = [1];
-        i--;
-        // Backtrack, nick not followed by anything it turns out
       }
-      linesToGet.push(lineObj);
     }
+    else if(state == 'new') {
+      specs.push(spec);
+      spec = {nicks: [], regexes: [], lines: []};
+    }
+
+    else if(state == 'error') {
+      return {error: error};
+    }
+
+    // Get next state
+    result = intuitState(input, i, state);
+    i = result.index;
+    state = result.state;
+    if(result.error !== undefined) {
+      error = result.error;
+    }
+  }
+
+  specs.push(spec);
+
+  return _.filter(specs, function(spec) {
+    return (spec.nicks.length > 0) ||
+      (spec.regexes.length > 0) ||
+      (spec.lines.length > 0);
+  });
+}
+
+function parseNick(str, idx) {
+  var ret = "";
+
+  while(idx < str.length && str[idx] != ' ' && str[idx] != ',') {
+    ret += str[idx];
```

Comment:
I don't think that's more readable even if it's more functional and pretty and clever